### PR TITLE
[DUOS-662][risk=no] New V2 POST, PUT for Draft Data Access Requests

### PR DESCRIPTION
**Pull request by rushtong**
## Addresses
https://broadinstitute.atlassian.net/browse/DUOS-662
Also partially implements security risk issues found in https://broadinstitute.atlassian.net/browse/DUOS-442, specifically the `PUT` requests for dars and draft dars.

## Changes
* New v2 POST endpoint for draft DARs
* New v2 PUT endpoint for draft DARs
* Updated api docs
* Added external collaborators to DAR fields

### CODE REVIEW

File: src/main/resources/assets/api-docs.yaml [hunk 1]
```diff
@@ -3056,6 +3100,25 @@ definitions:
       updateStatus:
         type: boolean
         description: Indicates if a consent has been updated from orsp
+  Collaborator:
+    type: object
+    description: A Collaborator on a Data Access Request
+    properties:
+      approverStatus:
+        type: boolean
+        description: Collaborator approver/downloader status
```
Comment (JVThomas) [inline src/main/resources/assets/api-docs.yaml]: Can you add a note about how approverStatus is relevant to lab staff and not to internal collaborators? The nuance behind approverStatus is not obvious from the name alone.

File: src/main/resources/assets/api-docs.yaml [hunk 2]
```diff
@@ -3205,6 +3268,18 @@ definitions:
       sharingDistribution:
         type: boolean
         description: Study requires sharing distribution
+      internalCollaborators:
+        type: array
+        description: Array of ontology terms
```
Comment (JVThomas) [inline src/main/resources/assets/api-docs.yaml]: You need to update the description here

File: src/main/resources/assets/api-docs.yaml [hunk 3]
```diff
@@ -3205,6 +3268,18 @@ definitions:
...
+        items:
+          type: object
+          $ref: '#/definitions/Collaborator'
+      labCollaborators:
+        type: array
+        description: Array of ontology terms
```
Comment (JVThomas) [inline src/main/resources/assets/api-docs.yaml]: Same issue with description as internalCollaborators

File: src/main/resources/assets/api-docs.yaml (hunk 1 above)
Comment (rushtong) [reply src/main/resources/assets/api-docs.yaml]: Yes, can do.

File: src/main/resources/assets/api-docs.yaml (hunk 2 above)
Comment (rushtong) [reply src/main/resources/assets/api-docs.yaml]: Good catch!

File: src/main/resources/assets/api-docs.yaml (hunk 3 above)
Comment (rushtong) [reply src/main/resources/assets/api-docs.yaml]: Good catch!

**Review by lindojong — COMMENTED** [review]
Tried out the endpoint locally using the example input data model and noticed two small inconsistencies between doc data model and actual data model: datasetIds as Array of String -> Array of Integer, ontologies as Array of String -> Array of OntologyEntry. Maybe it's worth updating the model definition for DataAccessRequest inside api-docs?

**Comment by rushtong** [comment]
Thank you @lindojong - it is worth updating the docs. Will do that now.

**Review by JVThomas — APPROVED** [review]
Dates look good. createDate is only defined on POST and is simply returned on the PUT request, and none of the dates have zeroed out times. Collaborators are saving properly as well.

**Comment by rushtong** [comment]
I'm going to add one more check to the PUTs to ensure the same user is making the update, then call it a day for this PR.

**Comment by JVThomas** [comment]
User validation looks like it's working. Separate issue: a user can send in a payload that changes the reference id value within the json, causing a mismatch between the column and the json attribute. Instead of simply accepting the user input, it'll be a good idea to just assign the value to the json manually server-side by reading the value from the db record itself.

**Comment by rushtong** [comment]
Nice validation catch, will correct that.

**Comment by rushtong** [comment]
@JVThomas Feedback applied, please have another look at your convenience!

**Comment by JVThomas** [comment]
Another nitpick: The CREATE/POST endpoint is reading the reference_id from the json payload and assigning it as a value for both the json and db column. The endpoint should be generating the value itself and assigning that value to those attributes.

**Comment by rushtong** [comment]
@JVThomas Another nice catch. Updated.

**Review by JVThomas — APPROVED** [review]
CREATE looks good, it ignores the user supplied json values for reference_id and instead applies its own generated value. PR as a whole looks good to me.

**Pull request merged**